### PR TITLE
eg25-manager: init at 0.4.6

### PR DESCRIPTION
**Pull request by Luflosi**
## Description of changes
https://gitlab.com/mobian1/eg25-manager
Manager daemon for the Quectel EG25 mobile broadband modem found on the Pine64 PinePhone and PinePhone Pro.
There was a previous attempt to add this package to Nixpkgs (#129229) but it was rejected because "This is very specific to the pinephone, I feel like it would make more sense to include in mobile-nixos." However I disagree with this reasoning.
- It is absolutely possible to run NixOS on the PinePhones using U-Boot or Tow-Boot without using Mobile NixOS.
- I also like to track all packages I maintain using Repology (https://repology.org/project/eg25-manager/versions) to see if any packages need updating but Mobile NixOS is not supported. Adding this Package to NixOS would allow me to better keep track of it.
- While Hydra does build Mobile NixOS and with it eg25-manager, it does not always use the latest Nixpkgs version. Looking at https://hydra.nixos.org/jobset/mobile-nixos/unstable one can see that currently the last evaluation was on 2024-07-11 which is definitely older than the latest unstable channel. I therefore have to build eg25-manager myself most of the time since I mostly use the latest Nixpkgs (unstable).
- Someone else also argued that eg25-manager should be upstreamed to Nixpkgs: https://github.com/mobile-nixos/mobile-nixos/issues/667. The original reason stated in the issue isn't valid since eg25-manager doesn't work with USB versions of this modem but @wentam commented "Personally I think it would make sense to have this in nixpkgs [...]".
- Last year I added eg25-manager to Mobile NixOS in https://github.com/mobile-nixos/mobile-nixos/pull/573 but I always felt that it should live in Nixpkgs instead.

The code in this PR is the code currently in Mobile NixOS with some minor cleanups.

Cc @Pacman99 @tomfitzhenry @kurnevsky @rvl @justinlovinger

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/eg/eg25-manager/package.nix [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  src = fetchFromGitLab {
+    owner = "mobian1";
+    repo = "eg25-manager";
+    rev = finalAttrs.version;
+    hash = "sha256-2JsdwK1ZOr7ljNHyuUMzVCpl+HV0C5sA5LAOkmELqag=";
+  };
+
+  postPatch = ''
+    substituteInPlace 'udev/80-modem-eg25.rules' \
+      --replace-fail '/bin/grep' '${gnugrep}/bin/grep'
```
Comment (AndersonTorres) [inline pkgs/by-name/eg/eg25-manager/package.nix:31]:
```suggestion
      --replace-fail '/bin/grep' '${lib.getExe gnugrep}'
```

File: pkgs/by-name/eg/eg25-manager/package.nix [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    glib
+    libgpiod_1 # Tracking issue for compatibility with libgpiod 2.0: https://gitlab.com/mobian1/eg25-manager/-/issues/45
+    libgudev
+    libusb1
+    modemmanager
+  ];
+
+  strictDeps = true;
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/by-name/eg/eg25-manager/package.nix:55]:
```suggestion
  meta = {
```

File: pkgs/by-name/eg/eg25-manager/package.nix [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    description = "Manager daemon for the Quectel EG25 mobile broadband modem found on the Pine64 PinePhone and PinePhone Pro";
+    homepage = "https://gitlab.com/mobian1/eg25-manager";
+    changelog = "https://gitlab.com/mobian1/eg25-manager/-/tags/${version}";
+    license = licenses.gpl3Plus;
+    platforms = platforms.linux;
+    mainProgram = "eg25-manager";
+    maintainers = with maintainers; [ Luflosi ];
```
Comment (AndersonTorres) [inline pkgs/by-name/eg/eg25-manager/package.nix:62]:
```suggestion
    license = lib.licenses.gpl3Plus;
    mainProgram = "eg25-manager";
    maintainers = with lib.maintainers; [ Luflosi ];
    platforms = lib.platforms.linux;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/hardware/network/eg25-manager.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (AndersonTorres) [inline nixos/modules/hardware/network/eg25-manager.nix:1]: Split this file in its own commit, with message 

```
nixos/eg25-manager: init
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Besides LGTM

**Comment by justinlovinger** [comment]
Also, note that Mobile NixOS is not fully compatible with NixOS. It uses an experimental kernel-builder that is incompatible with NixOS modules enabling kernel-features, <https://github.com/mobile-nixos/mobile-nixos/issues/715>. It is even incompatible with NixOS defaults since the latest release. There are good reasons to use NixOS on a mobile device without using Mobile NixOS.

I agree with moving this out of Mobile NixOS, but I wonder whether Nixpkgs or `nixos-hardware` is more appropriate for this. I am not advocating either way. What is the precedent?

**Comment by Luflosi** [comment]
> I agree with moving this out of Mobile NixOS, but I wonder whether Nixpkgs or `nixos-hardware` is more appropriate for this. I am not advocating either way. What is the precedent?

I think `nixos-hardware` is more for firmware, kernel forks, device-specific kernel parameters, kernel modules, CPU microcode and similar tweaks. There doesn't seem to be much precedent for adding entirely new packages to `nixos-hardware`, so I would prefer Nixpkgs itself for this.

**Comment by AndersonTorres** [comment]
Nixpkgs has basically no restrictions for adding things.
Further, ideally I believe nixos-hardware should be merged into the monorepo.

**Review by rvl — APPROVED** [review]
Excellent, thanks.
